### PR TITLE
Adding CPUID for AVX10.2

### PR DESCRIPTION
**Pull request by khushal1996**
### Overview
This PR tracks addition of CPUID for AVX10.2 to runtime. We are following the [spec doc](https://cdrdv2-public.intel.com/828965/361050-intel-avx10.2-spec.pdf) to add the CPUID for AVX10.2. Following things have been modified as a part of this PR

1. Adding a new ISA ```InstructionSet_Avx10v2``` and ```InstructionSet_Avx10v2_V512```
2. Adding a new environment variable ```EnableAvx10v2``` to JIT.
3. Handle enabling/disabling of AVX10 ISA in a careful manner. ```EnableAvx10v1``` will take priority over ```EnableAvx10v2``` i.e. if ```Avx10v1``` is disabled, ```Avx10v2``` will be disabled too.
4. Remove checks for CPUID for AVX10v1_V128 since bit 17 will imply V128 and V256 support i.e. minimum support on AVX10.
image

### Testing Correctness of Logic
Since we don't have the DMR hardware available right now, the testing for this PR was done on internal SDE. Following scenarios were tested on internal SDE

image

1. Add debug prints to check the value of ```cpufeatures```  after setting all the ISA flags.
image
image
image

2. Enable ```AVX10v1``` and ```AVX10v2``` by default and do the same. 
image

3. Disable ```Avx10v1``` and check if it disables ```Avx10v2```. (it should disable ```AVX10v2```)
image

4. Disable ```Avx10v2``` and verify that ```Avx10v1``` stays enabled.
image

### Testing JIT test suite
----
The testing was done using the Rex2 unit testing framework. All the changes made for testing CPUID changes are present in [this repo](https://github.com/khushal1996/runtime/tree/kcm-avx102-cpuid-sde-testing)
 
A brief overview --> 
- Unit tests ran to make sure no existing test is breaking
- SDE embedded in test framework
- Test ran with Intel internal SDE using ```dmr``` machine

**Step 1**: Run all tests without SDE
image

**Step 2**: Running all tests on the **current branch** with ```sde -dmr``` and ```DOTNET_EnableAVX10v2=0```
image

**Step 3**: Running all tests on the **current branch** with ```sde -dmr```
image

### Testing ASMDIFFS
----
Using ```jit-diff``` method, this testing step compares assembly between the ```main``` and the current branch and makes sure that we are not changing any existing scenarios.

Note - we cannot use superpmi here since we are adding a new CPUID and hence the ```jiteeversionguid``` has changed which would lead to wrong results from superpmi.

image

### CODE REVIEW

File: src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs
Comment (saucecontrol) [inline src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs:1964]: This generated code checks the incorrect top-level class name. Fix is in https://github.com/dotnet/runtime/pull/109210

Comment (khushal1996) [reply src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs:1964]: Thanks for the update. I see you PR is approved. Will wait for you to merge and rebase my PR accordingly.

**Comment by khushal1996** [comment]
@tannergooding Can you please help review this PR?

**Comment by tannergooding** [comment]
This has conflicts that need resolving now that #104637 is merged

**Comment by khushal1996** [comment]
@tannergooding  PR has been rebased and updated.

File: src/coreclr/jit/hwintrinsic.cpp
Comment (saucecontrol) [inline src/coreclr/jit/hwintrinsic.cpp:837]: APX_X64 is unused.  It would be better to remove it from InstructionSetDesc.txt than to add the entry for it here.

Comment (khushal1996) [reply src/coreclr/jit/hwintrinsic.cpp:837]: APX CPUID was enabled earlier today. Although it it unused, these are auto generated ISAs and they are added whenever we define a new ISA in ```InstructionSetDesc.txt```. I don't think removing ```APX_X64``` is an option here.

Comment (saucecontrol) [reply src/coreclr/jit/hwintrinsic.cpp:837]: It's added by this line, which shouldn't have been added since it's not needed.

https://github.com/dotnet/runtime/blob/30661f53e1a60d0d18d21488f4ab55e7e3621030/src/coreclr/tools/Common/JitInterface/ThunkGenerator/InstructionSetDesc.txt#L97

Removing that and re-generating will get rid of it everywhere.

Comment (tannergooding) [reply src/coreclr/jit/hwintrinsic.cpp:837]: It was standard practice, because that's what the tool was setup for. However, the logic for the generator was recently updated and so it shouldn't be required anymore if there is no actual need for it.

File: src/coreclr/vm/codeman.cpp
Comment (tannergooding) [inline src/coreclr/vm/codeman.cpp:1427]: nit: We normally don't nest the top level ISA checks to reduce the overall nesting and improve readability.

Instead, we rely on the `CPUCompileFlags.EnsureValidInstructionSetSupport` call down below to ensure they're consistent if some environment variable disabled a dependency.

**Comment by tannergooding** [comment]
I think we're missing some handling in:
* src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs
* src/coreclr/tools/Common/InstructionSetHelpers.cs
* src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/HardwareIntrinsicHelpers.Aot.cs

Since we don't have managed ISAs yet we don't need to handle:
* src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.NoX86Intrinsics.xml
* src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.x86.xml
* src/mono/mono/mini/simd-intrinsics.c

File: src/coreclr/jit/hwintrinsic.cpp
Comment (tannergooding) [reply src/coreclr/jit/hwintrinsic.cpp:837]: We want `AVX10v2_X64` because it has a managed API surface exposed. @saucecontrol was calling out `APX_X64` in particular, which is really unrelated to this PR and something to clean up independently

File: src/coreclr/vm/codeman.cpp
Comment (khushal1996) [reply src/coreclr/vm/codeman.cpp]: I have updated the PR accordingly.

**Comment by khushal1996** [comment]
> I think we're missing some handling in:
> 
> * src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs
> * src/coreclr/tools/Common/InstructionSetHelpers.cs
> * src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/HardwareIntrinsicHelpers.Aot.cs
> 
> Since we don't have managed ISAs yet we don't need to handle:
> 
> * src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.NoX86Intrinsics.xml
> * src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.x86.xml
> * src/mono/mono/mini/simd-intrinsics.c

Thanks for pointing me to this. Looks like this handling was added later to handle ```AVX10v1``` support. I have updated the handling for ```AVX10v2``` accordingly.

File: src/coreclr/jit/hwintrinsic.cpp
Comment (tannergooding) [reply src/coreclr/jit/hwintrinsic.cpp:837]: I'm fine with either adding the `hwintrinsicIsaRangeArray` for `APX_X64` here and then it being removed in a follow up -or- just removing `InstructionSet_APX_X64` in this PR directly so that the range doesn't need to be added at all.

Will leave it up to you on what you think is easiest/best.

**Review by tannergooding — APPROVED** [review]

**Comment by tannergooding** [comment]
CC, @dotnet/jit-contrib for secondary review

File: src/coreclr/jit/hwintrinsic.cpp
Comment (khushal1996) [reply src/coreclr/jit/hwintrinsic.cpp]: I have removed ```APX_X64``` ISA.

File: src/coreclr/scripts/out.txt
Comment (BruceForstall) [inline src/coreclr/scripts/out.txt:1]: Remove this file

File: src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs
Comment (BruceForstall) [inline src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs:379]: Is this backwards? It seems that if avx10v2_v512 is supported then avx10v1_v512 is also supported, but not vice-versa.

File: src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs
Comment (BruceForstall) [inline src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs:1250]: This seems odd: it says that if AVX10v1 is supported, then AVX10v2 is supported. Is that correct? I assume the OPPOSITE is true.

Comment (tannergooding) [reply src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs:1250]: This is part of `ExpandInstructionSetByReverseImplicationHelper`, which exists to allow NAOT to enable any dependent ISAs if the user just specifies `avx10v2` on the command line.

Comment (anthonycanino) [reply src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs:1250]: The opposite is true. @khushal1996 please correct this.

File: src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs
Comment (tannergooding) [reply src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs:379]: This one does look incorrect. I believe it should be `if (_supportedInstructionSets.Contains("avx10v2"))` so it matches the AVX10v1 handling just above.

Which is to say that if `Avx10v1` is supported and `AVX512` is supported, then `Avx10v1_V512` is supported.

File: src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs
Comment (tannergooding) [reply src/coreclr/tools/Common/JitInterface/CorInfoInstructionSet.cs:1250]: @anthonycanino this part is correct (and is notably autogenerated by the tool). Its part of `ExpandInstructionSetByReverseImplicationHelper`

File: src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs
Comment (khushal1996) [reply src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs:379]: I had deliberately kept this as is. By implication

implication        ,X86   ,AVX10v2              ,AVX10v1
implication        ,X86   ,AVX10v2_V512         ,AVX10v1_V512

And what bruce is saying is correct. If we have ```AVX10v2_V512```, then we would have ```AVX10v1_V512```.  Please correct me if I am understanding it wrong. From the design, we cannot have ```AVX10v1``` and have ```AVX10v2```.

Comment (khushal1996) [reply src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs]: Pushed this change.

Comment (khushal1996) [reply src/coreclr/tools/Common/Compiler/InstructionSetSupport.cs]: Makes sense. thanks for pointing this out @BruceForstall @tannergooding

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**
